src/invidious/helpers/extractors.cr

```diff
@@ -0,0 +1,562 @@
+  #
+  # "twoColumnBrowseResultsRenderer": {
+  #   {"tabs": [
+  #     {"tabRenderer":  {
+  #       "endpoint": {...}
+  #       "title": "Playlists",
+  #       "selected": true,
+  #       "content": {...},
+  #       ...
+  #     }}
+  #   ]}
+  # }]
+  #
+  module YouTubeTabs
+    def self.process(initial_data : Hash(String, JSON::Any))
+      if target = initial_data["twoColumnBrowseResultsRenderer"]?
+        self.extract(target)
+      end
+    end
+
+    private def self.extract(target)
+      raw_items = [] of JSON::Any
+      content = extract_selected_tab(target["tabs"])["content"]
+
+      content["sectionListRenderer"]["contents"].as_a.each do |renderer_container|
+        renderer_container_contents = renderer_container["itemSectionRenderer"]["contents"].as_a[0]
+
+        # Category extraction
+        if items_container = renderer_container_contents["shelfRenderer"]?
+          raw_items << renderer_container_contents
+          next
+        elsif items_container = renderer_container_contents["gridRenderer"]?
+        else
+          items_container = renderer_container_contents
+        end
+
+        items_container["items"].as_a.each do |item|
+          raw_items << item
+        end
+      end
+
+      return raw_items
+    end
+  end
+
+  # Extracts items from the InnerTube response for search results
+  #
+  # Search results are typically stored under "twoColumnSearchResultsRenderer"
+  # and is structured like this:
+  #
+  # "twoColumnSearchResultsRenderer": {
+  #   {"primaryContents": {
+  #     {"sectionListRenderer": {
+  #       "contents": [...],
+  #       ...,
+  #       "subMenu": {...},
+  #       "hideBottomSeparator": true,
+  #       "targetId": "search-feed"
+  #     }}
+  #   }}
+  # }
+  #
+  module SearchResults
+    def self.process(initial_data : Hash(String, JSON::Any))
+      if target = initial_data["twoColumnSearchResultsRenderer"]?
+        self.extract(target)
+      end
+    end
+
+    private def self.extract(target)
+      raw_items = [] of Array(JSON::Any)
+
+      target.dig("primaryContents", "sectionListRenderer", "contents").as_a.each do |node|
+        if node = node["itemSectionRenderer"]?
+          raw_items << node["contents"].as_a
+        end
+      end
+
+      return raw_items.flatten
+    end
+  end
+
+  # Extracts continuation items from a InnerTube response
+  #
+  # Continuation items (on YouTube) are items which are appended to the
+  # end of the page for continuous scrolling. As such, in many cases,
+  # the items are lacking information such as author or category title,
+  # since the original results has already rendered them on the top of the page.
+  #
+  # The way they are structured is too varied to be accurately written down here.
+  # However, they all eventually lead to an array of parsable items after traversing
+  # through the JSON structure.
+  module Continuation
+    def self.process(initial_data : Hash(String, JSON::Any))
```

Comment:
Now that I think about it, all youtube API functions should return a `JSON::Any` object, and not a `Hash`. It's a useless conversion that probably implies to copy many elements in memory.

Though, that will be something for another PR...